Added a function to get all resources in the open data website with one API call.

haritj01 opened the pull request:
Closes #82

haritj01 committed:
Added a function to get all resources in the open data website.

haritj01 committed:
Style code (GHA)

haritj01 committed:
Document package (GHA)

Moohan commented on R/get_all_resources.R:
```diff
@@ -0,0 +1,73 @@
+
+  out <- jsonlite::fromJSON(jsonlite::toJSON(out$resources), flatten = TRUE)
+
+  out <- data.table::rbindlist(out, use.names = TRUE, fill = TRUE)
```
This will create a new dependency on `{data.table}`. It's good to keep dependencies minimal wherever possible. We already have `{purrr}` though, and I think this will do the same thing: <https://purrr.tidyverse.org/reference/list_c.html>

```suggestion
  out <- purrr::list_rbind(out, use.names = TRUE, fill = TRUE)
```

Moohan reviewed (changes requested):
Great work finding `package_search`, seems exactly what is needed!

haritj01 committed:
Implemented requested changed from code review

haritj01 committed:
Style code (GHA)

haritj01 committed:
Document package (GHA)

Moohan reviewed (changes requested):
I was just having a quick look at this on my phone. So can't test anything.

I'm suspicious of the from_JSON / to_JSON pattern which seems redundant but I'll have to test and see the actual content to know for sure.

Also the way you build the query and add `rows=32000` doesn't look like how we do it in other places.

csillasch reviewed (commented):
Works great although errors when trying to use resource_contains - I`ve commented against the line I think is causing it.

I'd suggest naming the function something like list_all_resources or search_resources instead of get_all_resources, to mirror the list_datasets/list_resources functions and preserving the "get" prefix for functions that are returning the data within resources/functions rather than the metadata. Also wondering if we should call package_contains dataset_contains instead, as we've consistently referred to CKAN packages as datasets elsewhere to avoid confusion. What do you think?

haritj01 commented on R/get_all_resources.R:
```diff
@@ -0,0 +1,73 @@
+
+  if (!is.null(resource_contains)) {
+    data_tibble <- data_tibble[grepl(as.character(resource_contains), data_tibble$resource_name, ignore.case = TRUE), ]
+    if (nrow(out) == 0) {
```
Ah yes sorry my bad. I missed it while renaming the variables.

haritj01 committed:
Changed the function name to list_all_resources, fixed an error with package_contains and used cli_abort() for warnings

haritj01 committed:
Style code (GHA)

haritj01 committed:
Document package (GHA)

haritj01 committed:
Added tests

haritj01 committed:
Style code (GHA)

csillasch reviewed (changes requested):
Thanks for making the changes, all working well and tests look good too. 

I think we should also replace "package" with "dataset" in the arguments & returned columns for consistency - we opted to use "dataset" instead of "package" in phsopendata so far to avoid confusion with R packages (even though this deviates from CKAN terminology we felt this would be clearer for users of phsopendata).

haritj01 committed:
Replaced "package" with "dataset"

haritj01 committed:
Document package (GHA)

haritj01 committed:
Replaced "package" with "dataset" in tests.

haritj01 committed:
Style code (GHA)

Moohan committed:
Fix / improve docs

Moohan committed:
Simplify error messages

Moohan committed:
Build the query as a list

Moohan committed:
Merge branch 'master' into get_all_resources_hj

Moohan committed:
Re-build README.Rmd

Moohan committed:
Add a NEWS item

Moohan committed:
Add Haritha Jagadeesh as a contributor

haritj01 committed:
Added a link to let the user know what a regular expression is

haritj01 committed:
Added a new function for just the query in list_all_resources

haritj01 committed:
updated the documentation

haritj01 committed:
Style code (GHA)

haritj01 committed:
Merge pull request #98 from Public-Health-Scotland/haritha_as_contributor

Add Haritha Jagadeesh as a contributor

Moohan committed:
Change documentation wording 'packages' -> 'dataset'

Moohan committed:
Improve, fix and simplify the tests

Moohan committed:
Normalise bad inputs

Moohan committed:
Error on non character input

Moohan committed:
Error on invalid regular expressions

Moohan committed:
Change warning message wording

Moohan committed:
Style code (GHA)

Moohan committed:
Rewrite the 'query' using `{purrr}` to operate on the list

This code is still quite complicated, but I think it is a bit easier to read than what we had before? I definitely think it should be easier to modify, for example, to add additional variables.

I discovered purrr::chuck and purrr::pluck, which I think we should use elsewhere in the package. They have nice syntax and good behaviour when objects you expect don't exist in the list.

The gist of how this works now is: Make a tibble for each dataset, then stick them all back together!

I also parsed the modified date into a POSIXct date type, as this matches what we do elsewhere.

Moohan commented:
@haritj01 I opened another PR with a rewrite of the query #100 using purrr. See what you think.

I've also committed a few changes to this branch, adding test and error / param checking.

I had the thought that this makes `list_resources' obsolete (and to a certain extent `list_datasets` too. Can you have a think, and we can chat!

Moohan committed:
Merge branch 'main' into get_all_resources_hj

Moohan committed:
Merge branch 'main' into get_all_resources_hj

haritj01 committed:
Rewrite the 'query' using `{purrr}` to operate on the list (#100)

haritj01 committed:
Both list_datasets() and list_resources() functions are now formally superseded by list_all_resources().
Documentation updated with lifecycle badges and guidance,
and deprecate_warn() messages added for clarity.

haritj01 committed:
Style code (GHA)

haritj01 committed:
Document package (GHA)

sbarnesphs committed:
Remove deprecate_warn() from superseded functions. (#107)

See:
https://lifecycle.r-lib.org/articles/communicate.html#supersede-a-function

Moohan committed:
Update deprecated function usage for dataset_name

Moohan committed:
Document package (GHA)

Moohan committed:
Merge branch 'main' into get_all_resources_hj

Moohan committed:
Move and rename `list_resources_query`

Moohan committed:
Increment version number to 1.0.3.9000

Moohan committed:
Update the NEWS item.

Moohan committed:
Include `dataset_title` in the query extract

Moohan committed:
Expand regex searching to split search terms by spaces

Moohan committed:
Add tests for new searching

Moohan committed:
Add simon as ctb (#109)

Add Simon as a contributor

csillasch commented on R/list_resources.R:
```diff
@@ -1,48 +1,222 @@
-#' `list_resources()` returns all of the resources associated
-#' with a dataset
+#' @description Provides an overview of all resources available from
+#'  [opendata.nhs.scot](https://www.opendata.nhs.scot/), with the option to limit results based on both package
```
```suggestion
#'  [opendata.nhs.scot](https://www.opendata.nhs.scot/), with the option to limit results based on both dataset
```

csillasch commented on R/list_resources.R:
```diff
@@ -1,48 +1,222 @@
 #'
-#' @inheritParams get_dataset
+#' @param dataset_contains A character string containing an expression to be
+#'  used as search criteria against the dataset 'title' field.
```
should we specify the search is on both dataset title and dataset name or is that more confusing than useful?
```suggestion
#' @param dataset_contains A character string containing an expression to be
#'  used as search criteria against the dataset name.
```

csillasch commented on NEWS.md:
```diff
@@ -1,3 +1,7 @@
+# phsopendata (development version)
+
+- [`list_resources()`](https://public-health-scotland.github.io/phsopendata/reference/list_resources.html) has been upgraded. Instead of just returning all resources from a dataset (given the exact title), it can now search both resources and dataset titles and will list all resources that match the search!
```
```suggestion
- [`list_resources()`](https://public-health-scotland.github.io/phsopendata/reference/list_resources.html) has been upgraded. Instead of just returning all resources from a dataset (given the exact dataset name), it can now search both resources and dataset titles and will list all resources that match the search!
```

csillasch reviewed (commented):
Couple suggestions to keep distinction between dataset name and title clearer

Moohan committed:
Improve doc wording

Co-authored-by: Csilla Scharle <59616009+csillasch@users.noreply.github.com>

haritj01 committed:
Document package (GHA)

Moohan committed:
Update NEWS.md

Co-authored-by: Csilla Scharle <59616009+csillasch@users.noreply.github.com>

Moohan committed:
Simplify datase_contains description

Co-authored-by: Csilla Scharle <59616009+csillasch@users.noreply.github.com>

Moohan committed:
Document package (GHA)

Moohan committed:
Merge branch 'main' into get_all_resources_hj
